### DISCUSSION TITLE
How to *not* set sys.excepthook?

### DISCUSSION THREAD

**Discussion by charmoniumQ** [fastapi/typer#1024] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer
import rich.traceback
rich.traceback.install(
    show_locals=False,
    # other awesome configs here, not just show_locals=False
    suppress=[
        # e.g.,
        #numpy.__file__,
    ],
)

app = typer.Typer()

@app.command()
def main():
    x = 2
    raise ValueError()

app()
# shows traceback WITH locals, even though I installed rich without locals.
# Yes, I know pretty_exceptions_show_locals exists, but locals are just an easily-demo-able representative for all of my Rich traceback configuration
```

### Description

It seems that Typer overwrites the `sys.excepthook` [here](https://github.com/fastapi/typer/blob/cc6623c34f6d9987e723b4663c0c903b51e2063c/typer/main.py#L318) and it's impossible to disable that functionality.

This is important because I have a nice `rich.traceback` configuration that suppresses stackframes, customizes width, etc. Typer calls `rich.traceback`, but lacks these customization options I want set.

Even passing `pretty_exceptions_short=False` still sets the `sys.excepthook` to the (ugly) stdlib excepthook [here](https://github.com/fastapi/typer/blob/cc6623c34f6d9987e723b4663c0c903b51e2063c/typer/main.py#L75). Arguably, that flag or perhaps a new flag should just avoid setting `sys.excepthook` on line 318.

I can implement this in a PR, but I was wondering if it would be worth it for others, or if that is a feature that would be accepted.

[rich_markup_mode=None](https://github.com/fastapi/typer/pull/859) is not right either; it disables _all_ of Rich markup; I just want Typer to not overwrite the `sys.excepthook`.

### Operating System

Linux

### Operating System Details

NixOS

### Typer Version

0.12.3

### Python Version

Python 3.12.5

**Comment by djachenko** [comment]
I've done some debugging and found I can capture `sys.excepthook` before `app()`  and put it  back at app callback, but it's still quite dirty hack.

**Comment by DhruvalAnandkar** [comment]
Hi @charmoniumQ, and thanks for the debugging, @djachenko - that capture/restore trick makes sense as a workaround, and it actually points at exactly what's going on under the hood.

I dug into Typer's source (main.py) to see exactly what's happening. Two things combine to cause this:

1st---> Typer only saves the ORIGINAL sys.excepthook once at import time - literally the first time `import typer` runs anywhere in your process not when app() is called.
2nd---> Every time you call app() the Typer unconditionally overwrites sys.excepthook with its own handler so that handler only falls back to the saved original when pretty_exceptions_enable=False.

Put those together and your example has the ordering backwards: import typer runs first so Typer captures the plain stdlib hook as its fallback - before your rich.traceback.install() call ever runs. That's exactly why you saw the stdlib traceback instead of your Rich config even when trying to disable Typer's own renderer.

The fix is to flip the order so Typer's snapshot captures your Rich hook instead:

```python
import rich.traceback

rich.traceback.install(
    show_locals=False,
    suppress=[
        # your config here
    ],
)

import typer # import typer AFTER installing your rich hook

app = typer.Typer(pretty_exceptions_enable=False)

@app.command()
def main():
    x = 2
    raise ValueError()

if __name__ == "__main__":
    app()
```

With pretty_exceptions_enable=False, Typer's hook falls straight through to the "original" hook it saved - which is now your fully-configured Rich installer, locals/suppress/width and all, instead of stdlib or Typer's own version.

One caveat worth knowing: this only works if nothing else in your dependency chain imports typer before your script does. If some other library imports it first, the snapshot happens then instead, and you'd need @djachenko's capture/restore approach as a fallback for that case.

Also worth linking: #508 is tracking this exact friction as an open issue, if you want a supported programmatic override down the line rather than relying on import order.

Hope this gets you the traceback config you were after! Let me know if the reordering works on your end. (And if this clears things up, I'd really appreciate it if you could mark this as the answer! )
